### DISCUSSION TITLE
[Intellisense] Support for classNames package

### DISCUSSION THREAD

**Discussion by martinlehoux** [tailwindlabs/tailwindcss#10254] [Ideas]
Hi there!

I think `classNames` is a pretty common package used with TailwindCSS. We currently use the VSCode Tailwind Intellisense extension, and the ESLint Tailwindcss plugin.

We wanted to get both ESLint and Intellisense on our tailwind classes, but these can be defined in many places.

The follwing snippet show the kind of thing I would love to see support in Tailwind Intellisense. ESLint plugin supports it easily with `calles: ["classNames"]`

```ts
const rightSectionItemContainerUtilities = classNames(
    'tpy-headline-6 hidden cursor-pointer lg:block',
    isNavbarTransparent ? 'text-base-white hover:text-primary-300' : 'text-neutral-800 hover:text-primary-500',
  );
```

I have been using `experimental.classRegex` to add some support, but it only works on simple cases (one single string in `classNames` call).

```
"experimental.classRegex": ["classNames\\('(.*)'\\)"]
```

Do you have any idea how we could improve that? I'd be happy to help improve this support

**Comment by wongjn** [comment]
You could look at changing your setting so it capture more inside the `classNames()` call, something like:
```json
{
  "tailwindCSS.experimental.classRegex": [
    ["classNames\\(([^)])+\\)", "'([^']*)'"],
  ],
}
```
Or if you wanted to support the ways that strings can be quoted in JavaScript:
```json
{
  "tailwindCSS.experimental.classRegex": [
    ["classNames\\(([^)])+\\)", "['\"`]([^'\"`]*)['\"`]"],
  ],
}
```
Though this could produce unexpected results with nested mixture of quote characters.

**Reply by martinlehoux** [reply]
Thanks a lot! I guess the solution was already available :+1: 

I havn't tested on all edge cases, but this fixed version does the job

```json
{
  "tailwindCSS.experimental.classRegex": [["classNames\\(([^)]+)\\)", "'([^']*)'"]],
}
```

**Reply by corydeppen** [reply]
I've been using something similar but it fails when I need to use `calc` for an arbitrary value (e.g. `cx("w-[calc(2*2rem)]")`). This works as expected, though - `cx("w-[2rem]")`. Does anyone happen to have a regex they use that works when `calc` is used?
